File: articles/what-is-systems-thinking/python/scenario_comparison.py
"""
Scenario comparison for a simplified institutional system.

The model is intentionally simple and synthetic.
"""

from __future__ import annotations

import csv
from pathlib import Path


DATA_DIR = Path(__file__).resolve().parents[1] / "data"
SCENARIO_PATH = DATA_DIR / "synthetic_scenarios.csv"


def load_scenarios() -> list[dict[str, str]]:
    with SCENARIO_PATH.open(newline="", encoding="utf-8") as handle:
        return list(csv.DictReader(handle))


def run_scenario(scenario: dict[str, str], periods: int = 12) -> dict[str, float | str]:
    trust = 62.0
    capacity = 55.0
    demand = 70.0
    delay = 14.0
    resilience = 40.0

    resource_multiplier = float(scenario["resource_multiplier"])
    demand_multiplier = float(scenario["demand_multiplier"])
    delay_multiplier = float(scenario["delay_multiplier"])
    shock_period = int(scenario["shock_period"])
    shock_size = float(scenario["shock_size"])

    for period in range(1, periods + 1):
        current_demand = demand * demand_multiplier
        if shock_period and period >= shock_period:
            current_demand += shock_size
        capacity += (resource_multiplier * 2.0) - (current_demand / 100.0)
        delay = max(1.0, delay * delay_multiplier + (current_demand - capacity) / 80.0)
        trust += (capacity / 100.0) - (delay / 20.0)
        resilience += (capacity / 120.0) - (current_demand / 150.0)

    return {
        "scenario_id": scenario["scenario_id"],
        "scenario_name": scenario["scenario_name"],
        "final_trust": round(trust, 2),
        "final_capacity": round(capacity, 2),
        "final_delay": round(delay, 2),
        "final_resilience": round(resilience, 2),
    }


def main() -> None:
    print("scenario_id,scenario_name,final_trust,final_capacity,final_delay,final_resilience")
    for scenario in load_scenarios():
        result = run_scenario(scenario)
        print(
            f"{result['scenario_id']},{result['scenario_name']},"
            f"{result['final_trust']},{result['final_capacity']},"
            f"{result['final_delay']},{result['final_resilience']}"
        )


if __name__ == "__main__":
    main()

File: _archive/pre-professional-script-upgrade/20260531-194442/what-is-systems-thinking/r/scenario_comparison.R
# Scenario comparison for "What Is Systems Thinking?"

get_script_dir <- function() {
  args <- commandArgs(trailingOnly = FALSE)
  file_arg <- grep("^--file=", args, value = TRUE)
  if (length(file_arg) > 0) {
    return(dirname(normalizePath(sub("^--file=", "", file_arg))))
  }
  return(getwd())
}

script_dir <- get_script_dir()
article_dir <- dirname(script_dir)
scenario_path <- file.path(article_dir, "data", "synthetic_scenarios.csv")
scenarios <- read.csv(scenario_path)

run_scenario <- function(row, periods = 12) {
  trust <- 62
  capacity <- 55
  demand <- 70
  delay <- 14
  resilience <- 40

  for (period in seq_len(periods)) {
    current_demand <- demand * row$demand_multiplier
    if (row$shock_period > 0 && period >= row$shock_period) {
      current_demand <- current_demand + row$shock_size
    }
    capacity <- capacity + (row$resource_multiplier * 2.0) - (current_demand / 100.0)
    delay <- max(1.0, delay * row$delay_multiplier + (current_demand - capacity) / 80.0)
    trust <- trust + (capacity / 100.0) - (delay / 20.0)
    resilience <- resilience + (capacity / 120.0) - (current_demand / 150.0)
  }

  data.frame(
    scenario_id = row$scenario_id,
    scenario_name = row$scenario_name,
    final_trust = round(trust, 2),
    final_capacity = round(capacity, 2),
    final_delay = round(delay, 2),
    final_resilience = round(resilience, 2)
  )
}

results <- do.call(rbind, lapply(seq_len(nrow(scenarios)), function(i) run_scenario(scenarios[i, ])))
print(results)
